pkg/dns/dns.go

```diff
@@ -103,29 +76,105 @@ func newHandler(truncate, IPv6 bool, hosts map[string]string, upstreamServers []
 		{Net: "tcp"},
 	}
 	h := &Handler{
-		truncate:     truncate,
+		truncate:     opts.TruncateReply,
 		clientConfig: cc,
 		clients:      clients,
-		IPv6:         IPv6,
-		cname:        make(map[string]string),
-		ip:           make(map[string]net.IP),
+		ipv6:         opts.IPv6,
+		cnameToHost:  make(map[string]string),
+		hostToIP:     make(map[string]net.IP),
 	}
-	for host, address := range hosts {
+	for host, address := range opts.StaticHosts {
 		if ip := net.ParseIP(address); ip != nil {
-			h.ip[Cname(host)] = ip
+			h.hostToIP[Cname(host)] = ip
 		} else {
-			h.cname[Cname(host)] = Cname(address)
+			h.cnameToHost[Cname(host)] = Cname(address)
 		}
 	}
 	return h, nil
 }
 
+func (h *Handler) ServeDNS(w dns.ResponseWriter, req *dns.Msg) {
+	switch req.Opcode {
+	case dns.OpcodeQuery:
+		h.handleQuery(w, req)
+	default:
+		h.handleDefault(w, req)
+	}
+}
+
+func Start(opts *ServerOptions) (*Server, error) {
+	server := &Server{}
+	if opts.UDPPort > 0 {
+		udpOptions := opts
```

Comment:
Since `opts` is a pointer, this just makes a new pointer (and not a copy of the struct).  You never notice this because we clobber `TruncateReply` on the next call, and then never look at the struct again anyway.